modules/planning/math/curve1d/piecewise_quintic_spiral_path.cc

```diff
@@ -0,0 +1,64 @@
+/******************************************************************************
+ * Copyright 2018 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+/**
+ * @file
+ **/
+
+#include "modules/planning/math/curve1d/piecewise_quintic_spiral_path.h"
+
+namespace apollo {
+namespace planning {
+
+PiecewiseQuinticSpiralPath::PiecewiseQuinticSpiralPath() {
+  accumulated_s_.push_back(0.0);
+}
+
+void PiecewiseQuinticSpiralPath::Append(QuinticSpiralPath spiral_path) {
+  double s = spiral_path.ParamLength() + accumulated_s_.back();
+  accumulated_s_.push_back(s);
+
+  pieces_.push_back(std::move(spiral_path));
+}
+
+double PiecewiseQuinticSpiralPath::Evaluate(const std::uint32_t order,
+    const double s) const {
+  auto index = LocatePiece(s);
+  return pieces_[index].Evaluate(order, s - accumulated_s_[index]);
+}
+
+double PiecewiseQuinticSpiralPath::DeriveKappaS(const double s) const {
+  auto index = LocatePiece(s);
+  const auto& piece = pieces_[index];
+  double ratio = (s - accumulated_s_[index]) / piece.ParamLength();
+  return piece.DeriveKappaDerivative(QuinticSpiralPath::DELTA_S, ratio);
+}
+
+std::size_t PiecewiseQuinticSpiralPath::LocatePiece(const double s) const {
+  CHECK(s >= accumulated_s_.front() && s <= accumulated_s_.back());
```

Comment:
it is better to tolerate this or return an error code.

Reply:
The class is dedicated for special purpose. The situation must not happen. So it is OK to have a strong assertion here.